libs/wingc/src/type_check.rs

```diff
@@ -991,6 +995,9 @@ pub struct TypeChecker<'a> {
 	source_path: &'a Path,
 
 	pub diagnostics: RefCell<Diagnostics>,
+
+	in_json: bool,
```

Comment:
Originally this was my intention to have the Typechecker have a mutable field like: `is_json` though the discussion that arises here: https://github.com/winglang/wing/pull/1524#discussion_r1119555854

is that it removes clutter but is not as maintainable as passing context.

Reply:
I'd consider the `TypeChecker` class as the mutable state used for context by design. It's identical to `self` in the `Visit` pattern. This still means the AST itself should remain immutable. But we need some mutable state during visiting (look at all the existing fields in `TypeChecker`). So I think you're original intention was spot on.

Reply:
I'm ok with either. I think it's useful to know at least that we can refactor out some of the state/context if necessary. (FWIW sometimes, depending on what you are trying to do in Rust, refactoring out structs into smaller pieces is actually needed in order to satisfy various borrow checking requirements -- check this article for an example: https://smallcultfollowing.com/babysteps/blog/2018/11/01/after-nll-interprocedural-conflicts/)

Reply:
Yea as well Im not strongly opinionated here so either way is fine with me.

Reply:
nice read, i see how we can decide to factor out some `Context` struct and still store it in `TypeChecker` if we need to in the future, without passing it as an arg all over the place.